### PR TITLE
added runtime migration warnings for numeric strings and string-number comparison

### PR DESCRIPTION
**Pull request by i582**
PHP 8 has changed the logic for comparing numbers and strings, as well as changes in the concept of a numeric string.

This PR adds the ability to enable warnings that the result in PHP 8 will differ with the result in PHP 7.

Added `set_migration_php8_warning` function to enable and disable. It takes a bitset, where the first bit is responsible for warning for string-number comparison ([RFC](https://wiki.php.net/rfc/string_to_number_comparison)), and the second, for changes in numeric strings ([RFC](https://wiki.php.net/rfc/saner-numeric-strings)).

Example of use:

```php
<?php
$b = 0;

set_migration_php8_warning(0b001);

if ($b == "") {}        // warning
if ("100" == "100 ") {} // no warning

set_migration_php8_warning(0b010);

if ($b == "") {}        // no warning
if ("100" == "100 ") {} // warning

set_migration_php8_warning(0b000);

if ($b == "") {}        // no warning
if ("100" == "100 ") {} // no warning
```

At the end of each request, the warnings are automatically disabled.

## String and number comparison

The following table shows how the result of some simple comparisons changes (or doesn't change):

| Comparison      | Before | After   |
| --------------- | ------ | ------- |
| `0 == "0"`      | `true` | `true`  |
| `0 == "0.0"`    | `true` | `true`  |
| `0 == "foo"`    | `true` | `false` |
| `0 == ""`       | `true` | `false` |
| `42 == "   42"` | `true` | `true`  |
| `42 == "42foo"` | `true` | `false` |

See full table in tests: https://github.com/VKCOM/kphp/blob/d5a76ad890e46244e94a5e3003212383e351ca14/tests/cpp/runtime/number-string-comparison.cpp#L21

### Note
We are already comparing infinities as in PHP 8.
See [Special values](https://wiki.php.net/rfc/string_to_number_comparison#special_values) part of RFC.

## Changes in numeric strings

In PHP 8 numeric strings with trailing spaces are also considered valid numeric strings.

`string::is_numeric()` now uses `try_to_float` internally to handle the new behavior correctly.

Before

```php
42 == "42"; // true
42 == "   42"; // true
42 == "42   "; // false
```

After

```php
42 == "42"; // true
42 == "   42"; // true
42 == "42   "; // true
```

See https://github.com/VKCOM/kphp-polyfills/pull/25
#290

### CODE REVIEW

File: tests/phpt/migration_php8/006_is_numeric.php [hunk 1]
```diff
@@ -0,0 +1,13 @@
+@kphp_runtime_should_warn
+
```
Comment (DrDet) [inline tests/phpt/migration_php8/006_is_numeric.php:2]: It's better to specify warning message here

File: tests/phpt/migration_php8/006_is_numeric.php [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+$leading = " 100";
+$trailing = "100 ";
+$normal = "100";
+
+set_migration_php8_warning(0b010);
+
+var_dump(is_numeric($leading));
+var_dump(is_numeric($trailing)); // warning
+var_dump(is_numeric($normal));
```
Comment (DrDet) [inline tests/phpt/migration_php8/006_is_numeric.php:13]: Also, add test with `is_numeric('123  f')`, please

File: tests/phpt/migration_php8/002_string_number_comparison.php [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+/Comparison \(operator <\) results in PHP 7 and PHP 8 are different for 0 and "foo" \(PHP7: false, PHP8: true\)/
+/Comparison \(operator <\) results in PHP 7 and PHP 8 are different for 42 and "42foo" \(PHP7: false, PHP8: true\)/
+/Comparison \(operator <\) results in PHP 7 and PHP 8 are different for "100foo" and 42 \(PHP7: false, PHP8: true\)/
+<?php
+
+set_migration_php8_warning(0b001);
+
+echo 0 < "foo";
+echo 42 < "42foo";
+echo 42 > "100foo";
```
Comment (DrDet) [inline tests/phpt/migration_php8/002_string_number_comparison.php:11]: Also add some tests with spaceship operator please

**Review by DrDet — CHANGES_REQUESTED** [review]

File: tests/phpt/migration_php8/006_is_numeric.php (hunk 2 above)
Comment (i582) [reply tests/phpt/migration_php8/006_is_numeric.php:18]: Done.

File: runtime/migration_php8.cpp [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+// Copyright (c) 2021 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#include "migration_php8.h"
+
+int show_migration_php8_warning = 0;
+
+bool f$set_migration_php8_warning(int mask) {
+  show_migration_php8_warning = mask;
+  return true;
```
[the remark this answers is not in the document]
Comment (i582) [reply runtime/migration_php8.cpp]: Yep, done.

File: runtime/string.inl [hunk 5]
```diff
@@ -697,7 +698,29 @@ inline bool string::try_to_int(int64_t *val) const {
...
+  }
+
+  while (end_ptr != p + size() - 1) {
+    if (*end_ptr != ' ') {
+      return false;
+    }
+    end_ptr++;
+  }
+
+  return true;
```
[the remark this answers is not in the document]
Comment (i582) [reply runtime/string.inl]: Fixed.

File: runtime/mixed.inl [hunk 6]
```diff
@@ -1673,8 +1675,109 @@ inline int64_t operator>>(const mixed &lhs, const mixed &rhs) {
...
 }
 
+template <typename Arg1, typename Arg2>
+inline const char *conversion_php_warning_string() {
+  return "";
+}
+
+template<>
+inline const char *conversion_php_warning_string<int64_t, string>() {
+  return "Comparison (operator <) results in PHP 7 and PHP 8 are different for %" SCNd64 " and \"%s\" (PHP7: %s, PHP8: %s)";
```
[the remark this answers is not in the document]
Comment (i582) [reply runtime/mixed.inl]: Fixed.

File: tests/phpt/migration_php8/006_is_numeric.php (hunk 1 above)
Comment (i582) [reply tests/phpt/migration_php8/006_is_numeric.php]: Yep, done.

File: tests/phpt/migration_php8/002_string_number_comparison.php (hunk 3 above)
Comment (i582) [reply tests/phpt/migration_php8/002_string_number_comparison.php:11]: Added.

**Review by DrDet — APPROVED** [review]

**Review by DrDet — APPROVED** [review]

File: functions.txt [hunk 7]
```diff
@@ -1371,3 +1371,5 @@ function zstd_compress(string $data, int $level = 3) ::: string | false;
 function zstd_uncompress(string $data) ::: string | false;
 function zstd_compress_dict(string $data, string $dict) ::: string | false;
 function zstd_uncompress_dict(string $data, string $dict) ::: string | false;
+
+function set_migration_php8_warning ($show ::: int) ::: void;
```
Comment (unserialize) [inline functions.txt:1375]: (int $mask) ::: void

File: runtime/comparison_operators.inl [hunk 8]
```diff
@@ -75,15 +77,56 @@ inline bool eq2(const string &lhs, bool rhs) {
   return eq2(rhs, lhs);
 }
 
+// see https://www.php.net/manual/en/migration80.incompatible.php#migration80.incompatible.core.string-number-comparision
+template <typename T>
+inline bool equal_number_string_as_php8(T lhs, const string &rhs) {
```
Comment (unserialize) [inline runtime/comparison_operators.inl:82]: eq2_number_string_as_php8

File: runtime/migration_php8.h [hunk 9]
```diff
@@ -0,0 +1,14 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2021 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#pragma once
+
+constexpr int STRING_COMPARISON_FLAG = 1 << 0;
```
Comment (unserialize) [inline runtime/migration_php8.h:7]: MIGRATION_PHP8_STRING_COMPARISON_FLAG

File: runtime/migration_php8.h [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+constexpr int STRING_TO_FLOAT_FLAG = 1 << 1;
```
Comment (unserialize) [inline runtime/migration_php8.h:8]: MIGRATION_PHP8_STRING_TO_FLOAT_FLAG

File: functions.txt (hunk 7 above)
Comment (i582) [reply functions.txt]: Fixed.

File: runtime/migration_php8.h (hunk 9 above)
Comment (i582) [reply runtime/migration_php8.h]: Fixed.

**Review by unserialize — APPROVED** [review]

**Review by unserialize — APPROVED** [review]

File: runtime/string.inl [hunk 11]
```diff
@@ -820,10 +819,113 @@ bool string::is_int() const {
   return php_is_int(p, size());
 }
 
+bool string::is_numeric_as_php8() const {
+  const char *s = c_str();
```
Comment (DrDet) [inline runtime/string.inl:823]: Let's make separate function like `is_numeric_impl` and get rid of copypaste?

Comment (unserialize) [reply runtime/string.inl:823]: It seems an excessive refactoring, since php7-implementation will be dropped in the near future

Comment (DrDet) [reply runtime/string.inl:823]: ok, seems sensible

**Review by DrDet — APPROVED** [review]

**Pull request merged**